feat(graph): Flexible Schema Extension to SpannerGraphStore

mtyin opened the pull request:
feat(graph): Extension to SpannerGraphStore

Many use cases don't have a clean pre-specified schema, hence this change adds support to generic flexible graph schema.

1) Allow a dynamic schema for nodes and edges:
  - both nodes and edges are stored in one table with dynamic label and properties (JSON)
2) Allow more than one (src.node.id, target.node.id, edge.type) by introducing a uuid as part of the key;
3) Prefix the table names with graph names to avoid conflicting.
4) Add json type support.

Fixes #<issue_number_goes_here> 🦕

mtyin committed:
Extension to SpannerGraphStore

- Allow a dynamic schema for nodes and edges:
  - nodes are stored in one table with dynamic label and properties (JSON)
- Allow more than one (src.node.id, target.node.id, edge.type) by
  introducing a uuid as part of the key;
- Prefix the table names with graph names to avoid conflicting.

mtyin committed:
Fix linters

mtyin committed:
Update notebook for illustrate flexible schema

mtyin committed:
Remove the uuid key

amullick-git commented on src/langchain_google_spanner/graph_store.py:
```diff
@@ -230,59 +297,221 @@ def from_edges(name: str, edges: List[Relationship]) -> ElementSchema:
+        """
+        if len(edges) == 0:
+            raise ValueError("The list of edges should not be empty")
+
```
Lot of the code here is very similar to that of node. Would there be some value in making this a common function ?

mtyin replied:
Did some refactoring by distilling the common logic. PTAL

mtyin committed:
Resolve comments

mtyin commented on src/langchain_google_spanner/graph_store.py:
```diff
@@ -230,59 +297,221 @@ def from_edges(name: str, edges: List[Relationship]) -> ElementSchema:
+        Parameters:
+        - name: name of the schema;
+        - graph_name: name of the graph;
+        - nodes: a non-empty list of edges.
```
Good catch, done!

mtyin commented on src/langchain_google_spanner/graph_store.py:
```diff
@@ -230,59 +297,221 @@ def from_edges(name: str, edges: List[Relationship]) -> ElementSchema:
+        edge.original_name = name
+        edge.name = edge.base_table_name
+        edge.kind = "EDGE"
+
```
Done

mtyin commented on src/langchain_google_spanner/graph_store.py:
```diff
@@ -230,59 +297,221 @@ def from_edges(name: str, edges: List[Relationship]) -> ElementSchema:
+        - ElementSchema: schema representation of the edges.
+        """
+        if len(edges) == 0:
+            raise ValueError("The list of edges should not be empty")
```
It actually won't happen in practice. This is just for safety, to ensure the invariance.

amullick-git reviewed (approved):
Looks good to me!

mtyin committed:
Avoid unnecessary overwrites based on the specified properties

Previously, we override all columns by schema;
with this change, we selectively write specified columns.

Note:
- this won't change the behavior if you always do a full blind write;

mtyin committed:
Merge branch 'main' into extension

averikitsch commented on src/langchain_google_spanner/graph_store.py:
```diff
@@ -189,39 +265,75 @@ def from_nodes(name: str, nodes: List[Node]) -> ElementSchema:
+    ) -> ElementSchema:
+        """Builds ElementSchema from a list of nodes.
+
+        Parameters:
```
Use the google style for method docstrings, see https://google.github.io/styleguide/pyguide.html#doc-function-args. 
Missed this last time, so multiple functions may need to be updated. Thanks!

mtyin replied:
Done!

averikitsch commented on docs/graph_store.ipynb:
```diff
@@ -49,96 +31,104 @@
+    }
+   ],
+   "source": [
+    "%pip install --upgrade --quiet langchain-google-spanner langchain-google-vertexai langchain-experimental json-repair wikipedia pyvis youtube-transcript-api"
```
is json-repair still needed?

mtyin replied:
Yes, it's still required by the langchain-experimental library.

averikitsch commented on docs/graph_store.ipynb:
```diff
@@ -163,112 +155,70 @@
-  {
-   "cell_type": "code",
    "source": [
     "# @title Get graph documents from Wikipedia & YouTube\n",
```
```suggestion
    "# Get graph documents from Wikipedia & YouTube\n",
```
"@title" is used for colab variable forms

averikitsch commented on docs/graph_store.ipynb:
```diff
@@ -491,423 +502,73 @@
+    "graph_store.cleanup()"
    ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": []
```
Extra unused cell

mtyin committed:
Address comments

mtyin committed:
Minor updates to the interface

mtyin committed:
Add agent info for usage tracking

averikitsch commented on src/langchain_google_spanner/graph_store.py:
```diff
@@ -27,11 +27,13 @@
 DEFAULT_DDL_TIMEOUT = 300
 NODE_KIND = "NODE"
 EDGE_KIND = "EDGE"
+USER_AGENT_GRAPH_STORE = "langchain-google-spanner-python:graph_store/" + __version__
```
```suggestion
USER_AGENT_GRAPH_STORE = "langchain-google-spanner-python:graphstore/" + __version__
```

averikitsch committed:
Update src/langchain_google_spanner/graph_store.py

averikitsch committed:
Merge branch 'main' into extension
